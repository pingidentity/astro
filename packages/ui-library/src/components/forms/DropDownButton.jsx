var React = require("react"),
    ReactDOM = require("react-dom"),
    cx = require("classnames"),
    callIfOutsideOfContainer = require("../../util/EventUtils.js").callIfOutsideOfContainer,
    _ = require("underscore");

/**
 * @callback DropDownButton~onSelectCallback
 * @param {object} selectedValue - selected value from menu
 */

/**
 * @callback DropDownButton~onToggleCallback
 * @param {boolean} isOpen - current open/closed status before click
 */

/**
 * @class DropDownButton
 * @desc Button which triggers drop down menu using list of provided options.
 *
 * @param {string} options - array of string labels to use as drop down options
 * @param {DropDownButton~onSelectCallback} onSelect - callback to be triggered when selection changed.
 *                                                     Will recieve selected value from options list.
 * @param {DropDownButton~onToggleCallback} [onToggle] - callback to be triggered when open/closed state changed.
 *                                Used only with stateless mode. Will receive current open status value.
 * @param {bool} [open=false] - boolean state of open/closed menu. Used only in stateless mode.
 * @param {string} [id] - optional id to pass to be used as data-id attribute value
 * @param {string} [className] - optional class to pass to style top level container
 * @param {string} [label="Add"] - label text for button
 * @param {string} [title] - menu title text
 * @param {bool} [controlled=false] - A boolean to enable the component to be externally managed.  True will relinquish
 *   control to the components owner.  False or not specified will cause the component to manage state internally
 *   but still execute the onToggle callback in case the owner is interested.
 *
 * @example
 *
 *      <DropDownButton title="Drop Down"
 *                      id="drop-down-menu"
 *                      onSelect={this._changeARule}
 *                      options={optionsMenu}/>
 *
 *      <DropDownButton controlled={true}
 *                      title="My menu"
 *                      id="drop-down-menu"
 *                      onSelect={this._changeARule}
 *                      open={this.state.menuOpen}
 *                      onToggle={this._toggleMenu}
 *                      options={
 *                          one: "One",
 *                          two: "Two",
 *                          three: "Three"} />
 */


module.exports = React.createClass({
    propTypes: {
        controlled: React.PropTypes.bool
    },

    getDefaultProps: function () {
        return {
            controlled: false
        };
    },

    render: function () {
        return (
            this.props.controlled ? <Stateless {...this.props} /> : <Stateful {...this.props} />);
    }
});


var Stateless = React.createClass({

    propTypes: {
        options: React.PropTypes.object.isRequired,
        onSelect: React.PropTypes.func.isRequired,
        onToggle: React.PropTypes.func.isRequired,
        label: React.PropTypes.string,
        id: React.PropTypes.string,
        className: React.PropTypes.string,
        open: React.PropTypes.bool.isRequired,
        title: React.PropTypes.string
    },

    /**
     * Triggered
     * @method DropDownButton#_toggle
     * @param {object} e The event
     * @private
     */
    _toggle: function (e) {
        e.stopPropagation();
        this.props.onToggle(this.props.open);
    },

    /**
     * On option selected
     * @method DropDownButton#_onOptionSelected
     * @param {value} value The value selected from drop down
     * @private
     */
    _onOptionSelected: function (value) {
        if (this.props.onSelect) {
            this.props.onSelect(value);
        }
    },

    _handleGlobalClick: function (e) {
        if (this.props.open) {
            callIfOutsideOfContainer(
                ReactDOM.findDOMNode(this.refs.menu),
                _.partial(this.props.onToggle, this.props.open),
                e
            );
        }
    },

    _handleGlobalKeyDown: function (e) {
        if (e.keyCode === 27 && this.props.open) {
            this.props.onToggle(this.props.open);
        }
    },

    getDefaultProps: function () {
        return {
            label: "Add",
            title: "",
            options: {},
            open: false
        };
    },

    componentDidMount: function () {
        window.addEventListener("click", this._handleGlobalClick);
        window.addEventListener("keydown", this._handleGlobalKeyDown);
    },

    componentWillUnmount: function () {
        window.removeEventListener("click", this._handleGlobalClick);
        window.removeEventListener("keydown", this._handleGlobalKeyDown);
    },

    render: function () {

        var that = this,
            styles = cx("input-menu-button", {
                open: this.props.open
            }),
            content = null;

        if (this.props.className) {
            styles = styles + " " + this.props.className;
        }

        if (this.props.open) {

            var optionNodes = _.map(this.props.options, function (value, key) {

                return (
                    <a data-id={key} onClick={_.partial(that._onOptionSelected, key)} key={key}>{value}</a>
                );
            });

            content = (
                <div className="menu" ref="menu">
                    <div className="description">{this.props.title}</div>
                    <div className="options">
                        {optionNodes}
                    </div>
                </div>
            );
        }

        return (
            <div className={styles} data-id={this.props.id}>
                <a data-id="action" className="add button inline" onClick={this._toggle}>{this.props.label}</a>
                {content}
            </div>
        );
    }
});


var Stateful = React.createClass({

    _toggle: function () {
        this.setState({
            open: !this.state.open
        });
    },

    _select: function (value) {
        this.setState({
            open: false
        }, function () {
            if (this.props.onSelect) {
                this.props.onSelect(value);

            }
        });
    },

    getInitialState: function () {
        return {
            open: false
        };
    },

    render: function () {
        return (
            <Stateless {...this.props} onToggle={this._toggle} onSelect={this._select} open={this.state.open}/>
        );
    }
});

