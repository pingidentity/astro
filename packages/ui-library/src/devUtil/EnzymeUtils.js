import { mount } from "enzyme";


export const snapshotDataIds = node => {
    const elements = node.find("[data-id]").hostNodes(); // Find only actual HTML elements

    const ids = elements.map(element => element.prop("data-id")).filter(item => item !== "");
    expect(ids).toMatchSnapshot();
};

export const mountSnapshotDataIds = node => snapshotDataIds(mount(node));
