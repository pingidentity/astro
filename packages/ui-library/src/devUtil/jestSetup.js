
// This is needed for testing the modal and DetailsTooltip component.
global.getSelection = jest.fn();
global.getSelection.mockReturnValue({
    toString: () => "",
});
global.document.createRange = () => ({
    setStart: () => {},
    setEnd: () => {},
    commonAncestorContainer: {
        nodeName: "BODY",
        ownerDocument: document,
    },
});
