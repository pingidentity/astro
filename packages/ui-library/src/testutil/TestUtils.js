/*eslint-disable valid-jsdoc*/

"use strict";

/**
 * @module util/TestUtils
 * @desc A collection of ReactJS test utils, to complement the ones provided by the
 * framework. This is a copy of the TestUtils in the library, just because this class
 * is used in the tests and should work with the same version of React used in the WP
 * and not in the library.
 */

var ReactDOM = require("react-dom");
var React = require("react");
var ReactTestUtils = require("react-addons-test-utils");
var _ = require("underscore");
var assign = require("object-assign");

var TestUtils = {

    /**
     * @alias module:util/TestUtils.scryRenderedDOMNodesWithDataId
     * @desc Return all DOM nodes children of the given tree with the specified dataId.
     *
     * @param {object} parent - the React component or DOM Node tree to search
     * @param {string} dataId - the value of the data-id attribute on the target elements
     * @return {array} - of DOM node objects which match the given criteria
     */
    scryRenderedDOMNodesWithDataId: function (parent, dataId) {
        // first check if the parent component is a DOM node
        if (ReactDOM.findDOMNode(parent) === null) {
            return [];
        }

        var parentNode =
            ReactTestUtils.isDOMComponent(parent) || ReactTestUtils.isCompositeComponent(parent)
                ? ReactDOM.findDOMNode(parent)
                : parent;

        var nodes;
        if (parentNode.getAttribute("data-id") === dataId) {
            nodes = [parentNode];
        }
        else {
            var selectorValue = dataId
                .replace(/\[/g, "\\[")
                .replace(/\]/g, "\\]")
                .replace(/'/g, "\\'");
            nodes = parentNode.querySelectorAll("[data-id=" + selectorValue + "]");
        }

        return nodes;
    },

    /**
     * @alias module:util/TestUtils.findRenderedDOMNodeWithDataId
     * @desc Return the single DOM node child of the given tree with the specified dataId.
     * If no node in the tree matches the criteria, null is returned.
     * If there are more than one node, an error is thrown.
     *
     * @param {object} parent - the React component or DOM Node tree to search
     * @param {string} dataId - the value of the data-id attribute on the target element
     * @return {object} - the DOM node object which matches the given criteria
     */
    findRenderedDOMNodeWithDataId: function (parent, dataId) {
        var nodes = this.scryRenderedDOMNodesWithDataId(parent, dataId);

        if (nodes.length > 1) {
            throw new Error("Found more than one DOM node with dataId: " +
                dataId + " in the given tree");
        }

        return nodes.length === 0 ? null : nodes[0];
    },

    /**
     * @alias module:util/TestUtils.scryRenderedDOMNodesWithTag
     * @desc Return all DOM nodes children of the given tree with the specified tag name.
     *
     * @param {object} parent - the React component or DOM Node tree to search
     * @param {string} tagName - the tag name on the target elements
     * @return {array} - array of DOM nodes object which match the given criteria
     */
    scryRenderedDOMNodesWithTag: function (parent, tagName) {
        // first check if the parent component is a DOM node
        if (ReactDOM.findDOMNode(parent) === null) {
            return [];
        }

        var parentNode =
            ReactTestUtils.isDOMComponent(parent) || ReactTestUtils.isCompositeComponent(parent)
                ? ReactDOM.findDOMNode(parent)
                : parent;

        var nodes;
        if (parentNode.tagName.toLowerCase() === tagName.toLowerCase()) {
            nodes = [parentNode];
        }
        else {
            nodes = parentNode.querySelectorAll(tagName);
        }

        return nodes;
    },

    /**
     * @alias module:util/TestUtils.findRenderedDOMNodeWithTag
     * @desc Return the single DOM node child of the given tree with the specified tag name.
     * If no node in the tree matches the criteria, null is returned.
     * If there are more than one node, an error is thrown.
     *
     * @param {ReactComponent} parent - the React component or DOM Node tree to search
     * @param {string} tagName - the tag name on the target element
     * @return {DOMNode} - the DOM node which matches the given criteria
     */
    findRenderedDOMNodeWithTag: function (parent, tagName) {
        var nodes = this.scryRenderedDOMNodesWithTag(parent, tagName);

        if (nodes.length > 1) {
            throw new Error("Found more than one DOM node with tag name: " +
                tagName + " in the given tree");
        }

        return nodes.length === 0 ? null : nodes[0];
    },

    /**
     * @alias module:util/TestUtils.scryRenderedDOMNodesWithName
     * @desc Return all DOM nodes children of the given tree with the specified name.
     *
     * @param {ReactComponent} parent - the React component or DOM Node tree to search
     * @param {string} name - the name on the target elements
     * @return {DOMNode[]} - all DOM nodes which match the given criteria
     */
    scryRenderedDOMNodesWithName: function (parent, name) {
        // first check if the parent component is a DOM node
        if (ReactDOM.findDOMNode(parent) === null) {
            return [];
        }

        var parentNode =
            ReactTestUtils.isDOMComponent(parent) || ReactTestUtils.isCompositeComponent(parent)
                ? ReactDOM.findDOMNode(parent)
                : parent;

        var nodes;
        if (parentNode.name === name) {
            nodes = [parentNode];
        }
        else {
            nodes = parentNode.querySelectorAll("[name='" + name.replace("'", "\\'") + "']");
        }

        return nodes;
    },

    /**
     * @alias module:util/TestUtils.findRenderedDOMNodeWithName
     * @desc Return the single DOM node child of the given tree with the specified name.
     * If no node in the tree matches the criteria, null is returned.
     * If there are more than one node, an error is thrown.
     *
     * @param {ReactComponent} parent - the React component or DOM Node tree to search
     * @param {string} name - the name on the target element
     * @return {DOMNode} - the DOM node which matches the given criteria
     */
    findRenderedDOMNodeWithName: function (parent, name) {
        var nodes = this.scryRenderedDOMNodesWithName(parent, name);

        if (nodes.length > 1) {
            throw new Error("Found more than one DOM node with name: " +
                name + " in the given tree");
        }

        return nodes.length === 0 ? null : nodes[0];
    },

    /**
     * @alias module:util/TestUtils.scryRenderedDOMNodesWithClass
     * @desc Return all DOM nodes children of the given tree with the specified CSS class name.
     *
     * @param {ReactComponent} parent - the React component or DOM Node tree to search
     * @param {string} className - the CSS class name on the target elements (or a concatenation of classes, separated by space)
     * @return {DOMNode[]} - all DOM nodes which match the given criteria
     */
    scryRenderedDOMNodesWithClass: function (parent, className) {
        // first check if the parent component is a DOM node
        if (ReactDOM.findDOMNode(parent) === null) {
            return [];
        }

        var parentNode = ReactTestUtils.isDOMComponent(parent) || ReactTestUtils.isCompositeComponent(parent)
            ? ReactDOM.findDOMNode(parent)
            : parent;

        var nodes;

        // first check if the root element has the provided classes;
        // if not, build a selector and look for any children

        var parentMatches = _.every(className.split(" "), function (name) {
            return (" " + parentNode.className + " ").indexOf(" " + name + " ") > -1;
        });
        if (parentMatches) {
            nodes = [parentNode];
        }
        else {
            var selector = _.reduce(className.split(" "), function (memo, name) {
                return memo + "." + name;
            }, "");
            nodes = parentNode.querySelectorAll(selector);
        }

        return nodes;
    },

    /**
     * @alias module:util/TestUtils.findRenderedDOMNodeWithClass
     * @desc Return the single DOM node child of the given tree with the specified CSS class name.
     * If no node in the tree matches the criteria, null is returned.
     * If there are more than one node, an error is thrown.
     *
     * @param {ReactComponent} parent - the React component or DOM Node tree to search
     * @param {string} className - the CSS class name on the target element
     * @return {DOMNode} - the DOM node which matches the given criteria
     */
    findRenderedDOMNodeWithClass: function (parent, className) {
        var nodes = this.scryRenderedDOMNodesWithClass(parent, className);

        if (nodes.length > 1) {
            throw new Error("Found more than one DOM node with class name: " +
                className + " in the given tree");
        }

        return nodes.length === 0 ? null : nodes[0];
    },

    /**
     * @alias module:util/TestUtils.scryRenderedComponentsWithType
     * @desc Finds all instances of components with type equal to `componentType`.
     * @param {object} root - a composite (ie. React) component
     * @param {object} componentType - the component type
     * @return {array} an array of all the matches.
     */
    scryRenderedComponentsWithType: function (root, componentType) {
        return ReactTestUtils.scryRenderedComponentsWithType(root, componentType);
    },

    /**
     * @alias module:util/TestUtils.findRenderedComponentWithType
     * @desc Same as `scryRenderedComponentsWithType` but expects there to be one result
     * and returns that one result, or throws exception if there is any other
     * number of matches besides one.
     * @param {object} root - a composite (ie. React) component
     * @param {object} componentType - the component type
     * @return {!ReactComponent} The one match.
     */
    findRenderedComponentWithType: function (root, componentType) {
        return ReactTestUtils.findRenderedComponentWithType(root, componentType);
    },

    /**
     * @alias module:util/TestUtils.captureGlobalListener
     * @desc Attaches proxy function for `window.addEvenListener` which will save reference to actual provided callback when
     * event type match filter. Useful for testing ReactJS component global (window) callbacks behaviour.
     *
     * @param {string} eventType event to save callback for. case insensitive.
     * @param {object} [obj=window] optional object to capture listener from. Like `document`. If not provided - window object
     * @returns {function} - when executed will trigger actual provided callback to `window.addEventListener`
     *
     * @example
     * //capture registered callback from reactjs control
     * var onKeyDown = TestUtils.captureGlobalListener("keyDown");
     *
     * onKeyDown({keyCode:13});  //later in test trigger global event
     *
     * @example
     * //capture registered callback from reactjs control on `document` object
     * var onKeyDown = TestUtils.captureGlobalListener("keyDown", document);
     *
     * onKeyDown({keyCode:13});  //later in test trigger global event
     */
    captureGlobalListener: function (eventType, obj) {

        var captured;

        var target = obj || window;

        target.addEventListener = function (event, listener) {
            if (event.toLowerCase() === eventType.toLowerCase()) {
                captured = listener;
            }
        };

        return function () {
            return captured.apply(this, arguments);
        };
    }

    /**
     * @desc Return all components in the supplied tree with the specified dataId.
     *
     * @param {ReactComponent} tree - the React component tree to search
     * @param {string} dataId - the value of the data-id attribute on the target elements
     * @return {ReactComponent[]} - all components which match the given criteria
     */
    /*
     scryRenderedDOMComponentsWithDataId: function (tree, dataId) {
     var components = ReactTestUtils.findAllInRenderedTree(tree, function (inst) {
     var instId = ReactDOM.findDOMNode(inst).getAttribute("data-id");
     return ReactTestUtils.isDOMComponent(inst) && (instId === dataId);
     });

     return components;
     },
     */

    /**
     * @desc Return the single component in the supplied tree with the specified dataId.
     * If no component in the tree matches the criteria, null is returned.
     * If there are more than one component, an error is thrown.
     *
     * @param {ReactComponent} tree - the React component tree to search
     * @param {string} dataId - the value of the data-id attribute on the target element
     * @return {ReactComponent} - the component which match the given criteria
     * @throws Will throw an Error if more than one element match the criteria
     */
    /*
     findRenderedDOMComponentWithDataId: function (tree, dataId) {
     var components = TestUtils.scryRenderedDOMNodesWithDataId(tree, dataId);

     if (components.length > 1) {
     throw new Error("Found more than one component with dataId: " +
     dataId + " in the given tree");
     }

     return components.length === 0 ? null : components[0];
     },
     */

    /**
     * @desc Return the single component in the supplied tree with the specified name.
     * If no component in the tree matches the criteria, null is returned.
     * If there are more than one component, an error is thrown.
     *
     * @param {ReactComponent} tree - the React component tree to search
     * @param {string} name - the value of the name attribute on the target element
     * @return {ReactComponent} - the component which match the given criteria
     * @throws Will throw an Error if more than one element match the criteria
     */
    /*
     findRenderedDOMComponentWithName: function (tree, name) {
     var components = ReactTestUtils.findAllInRenderedTree(tree, function (inst) {
     var instName = ReactDOM.findDOMNode(inst).name;
     return ReactTestUtils.isDOMComponent(inst) && (name === instName);
     });

     if (components.length > 1) {
     throw new Error("Found more than one component with name: " +
     name + " in the given tree");
     }

     return components.length === 0 ? null : components[0];
     }
     */
};

/**
 * @class UpdatePropsWrapper
 *
 * @desc
 * A simple wrapper to be used for unit testing, when you need to update
 * a property on a rendered component. Starting with React 0.14.x,
 * directly setting or replacing a prop on a rendered component is deprecated.
 * Instead, use this wrapper and call _setProps(newProps) on it
 * to have the wrapped component re-render with the new property(ies).
 * The ref attribute on the wrapper is set to "wrapper".
 *
 * @Example
 *  // initial render
 *  var component = ReactTestUtils.renderIntoDocument(
 *      <UpdatePropsWrapper type={If} test={true} />
 *  );
 *  // update the property through re-render
 *  component._setProps({test: false});
 *
 * @Example
 *  var component = ReactTestUtils.renderIntoDocument(
 *      <UpdatePropsWrapper type={TabbedSections}
 *              onSectionChange={jest.genMockFunction()}
 *              selectedIndex={-1}>
 *          <div data-id="section1" title="section 1">section 1</div>
 *          <div data-id="section2" title="section 2">section 2</div>
 *      </UpdatePropsWrapper>
 *  );
 *  component._setProps({ selectedIndex: 0 });
 *
 */
TestUtils.UpdatePropsWrapper = React.createClass({
    /**
     * Re-render the wrapped component using the new property.
     * Use this function to avoid setting the props directly on the component (anti pattern).
     * @param {object} props - new property(ies) to set
     */
    _setProps: function (props) {
        this.setState(props);
    },
    getInitialState: function () {
        return assign({ ref: "wrapper" }, this.props);
    },
    render: function () {
        return React.createElement(
            this.props.type, // this is the type of element to create, provided as prop
            this.state // this is the rest of props to set on the element
        );
    }
});

module.exports = TestUtils;
