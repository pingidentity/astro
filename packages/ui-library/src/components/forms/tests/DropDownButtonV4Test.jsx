window.__DEV__ = true;

jest.mock("popper.js");
jest.mock("react-portal");

jest.dontMock("../DropDownButton");
jest.dontMock("../../../util/EventUtils.js");

import React from "react";
import ReactDOM from "react-dom";
import { mountSnapshotDataIds } from "../../../devUtil/EnzymeUtils";
import ReactTestUtils from "react-dom/test-utils";
import TestUtils from "../../../testutil/TestUtils";
import DropDownButton from "../DropDownButton";
import { mount } from "enzyme";

describe("DropDownButton", function () {
    it("data-id's don't change", () => {
        mountSnapshotDataIds(
            <DropDownButton
                title="title"
            />
        );
    });


    it("renders closed in stateless mode", function () {
        const callback = jest.fn();

        let menu = {
            optionOne: "Option One",
            optionTwo: "Option Two"
        };

        const dropDownButtonComponent = TestUtils.renderInWrapper(
            <DropDownButton title="Test Drop Down"
                stateless={true}
                onValueChange={callback}
                options={menu}
                onToggle={jest.fn()}
                label="Add"
            />
        );

        // expect a single button to be rendered.
        const button = TestUtils.findRenderedDOMNodeWithDataId(dropDownButtonComponent, "drop-down-button");
        expect(button.textContent).toBe("Add");

        // expect no modal to be rendered.
        menu = TestUtils.scryRenderedDOMNodesWithDataId(dropDownButtonComponent, "menu");
        expect(menu.length).toEqual(0);
    });

    it("renders open in stateless mode", function () {
        const callback = jest.fn();

        let menu = {
            optionOne: "Option One",
            optionTwo: "Option Two"
        };

        const dropDownButtonComponent = TestUtils.renderInWrapper(
            <DropDownButton label="Test Drop Down"
                stateless={true}
                open={true}
                onValueChange={callback}
                options={menu}
                onToggle={jest.fn()} />
        );

        //expect menu to be rendered
        menu = TestUtils.findRenderedDOMNodeWithDataId(dropDownButtonComponent, "menu");

        const as = TestUtils.scryRenderedDOMNodesWithTag(dropDownButtonComponent, "a");

        // make sure <a>s in a list, +1 length for button
        expect(as.length).toBe(3);
        expect(as[1].textContent).toEqual("Option One");
    });

    it("renders extra css classes", function () {
        const callback = jest.fn();

        const menu = {
            optionOne: "Option One",
            optionTwo: "Option Two"
        };

        const dropDownButtonComponent = TestUtils.renderInWrapper(
            <DropDownButton label="Test Drop Down"
                className="extra"
                stateless={true}
                open={true}
                onValueChange={callback}
                options={menu}
                onToggle={jest.fn()} />
        );

        //expect menu to be rendered
        const root = TestUtils.findRenderedDOMNodeWithDataId(dropDownButtonComponent, "drop-down-button");

        expect(root.getAttribute("class")).toContain("extra");
    });

    it("triggers callback on toggle in stateless mode", function () {
        const callback = jest.fn();

        const menu = {
            optionOne: "Option One",
            optionTwo: "Option Two"
        };

        const dropDownButtonComponent = TestUtils.renderInWrapper(
            <DropDownButton label="Test Drop Down"
                stateless={true}
                open={false}
                onValueChange={jest.fn()}
                onToggle={callback}
                options={menu} />
        );

        const button = TestUtils.findRenderedDOMNodeWithDataId(dropDownButtonComponent, "action");

        ReactTestUtils.Simulate.click(button);

        //make callback was triggered
        expect(callback).toBeCalledWith(false);
    });


    it("test default render", function () {

        const callback = jest.fn();

        let menu = {
            optionOne: "Option One",
            optionTwo: "Option Two",
            optionThree: "Option Three",
            optionFour: "Option Four",
            optionFive: "Option Five"
        };

        const dropDownButtonComponent = TestUtils.renderInWrapper(
            <DropDownButton label="Add" title="Test Drop Down"
                onValueChange={callback}
                options={menu} />
        );

        // expect a single button to be rendered.
        const button = TestUtils.findRenderedDOMNodeWithDataId(dropDownButtonComponent, "drop-down-button");
        expect(button.textContent).toBe("Add");

        // expect no modal to be rendered.
        menu = TestUtils.scryRenderedDOMNodesWithDataId(dropDownButtonComponent, "menu");
        expect(menu.length).toEqual(0);

    });

    it("initially closed menu opened on click", function () {

        const callback = jest.fn();

        let menu = {
            optionOne: "Option One",
            optionTwo: "Option Two",
            optionThree: "Option Three",
            optionFour: "Option Four",
            optionFive: "Option Five"
        };

        const dropDownButtonComponent = TestUtils.renderInWrapper(
            <DropDownButton label="Test Drop Down"
                onValueChange={callback}
                options={menu} />
        );

        // expect a single button to be rendered.
        const button = TestUtils.findRenderedDOMNodeWithDataId(dropDownButtonComponent, "action");

        // expect no menu to be rendered.
        const menus = TestUtils.scryRenderedDOMNodesWithDataId(dropDownButtonComponent, "menu");
        expect(menus.length).toEqual(0);

        ReactTestUtils.Simulate.click(button);

        // expect a menu to be rendered after click.
        menu = TestUtils.findRenderedDOMNodeWithDataId(dropDownButtonComponent, "menu");
        TestUtils.findRenderedDOMNodeWithClass(menu, "menu");

        const as = TestUtils.scryRenderedDOMNodesWithTag(dropDownButtonComponent, "a");

        // make sure <a>s in a list, +1 length for button
        expect(as.length).toBe(6);
        expect(as[1].textContent).toEqual("Option One");

        // click an option
        ReactTestUtils.Simulate.click(as[1]);

        // expect menu to have been hidden
        const noMenus = TestUtils.scryRenderedDOMNodesWithDataId(dropDownButtonComponent, "menu");
        expect(noMenus.length).toEqual(0);
    });

    it("will trigger callback on menu item click", function () {

        const callback = jest.fn();

        let menu = {
            optionOne: "Option One",
            optionTwo: "Option Two",
            optionThree: "Option Three",
            optionFour: "Option Four",
            optionFive: "Option Five"
        };

        const dropDownButtonComponent = TestUtils.renderInWrapper(
            <DropDownButton label="Test Drop Down"
                onValueChange={callback}
                options={menu} />
        );

        // expect a single button to be rendered.
        const button = TestUtils.findRenderedDOMNodeWithDataId(dropDownButtonComponent, "action");

        // expect no menu to be rendered.
        const menus = TestUtils.scryRenderedDOMNodesWithDataId(dropDownButtonComponent, "menu");
        expect(menus.length).toEqual(0);

        ReactTestUtils.Simulate.click(button);

        // expect a menu to be rendered after click.
        menu = TestUtils.findRenderedDOMNodeWithDataId(dropDownButtonComponent, "menu");
        TestUtils.findRenderedDOMNodeWithClass(menu, "menu");

        const as = TestUtils.scryRenderedDOMNodesWithTag(dropDownButtonComponent, "a");

        // make sure <a>s in a list, +1 length for button
        expect(as.length).toBe(6);
        expect(as[1].textContent).toEqual("Option One");

        // click an option
        ReactTestUtils.Simulate.click(as[1]);

        //make sure callback was triggered
        expect(callback).toBeCalled();
    });

    it("register global listeners on mount", function () {


        window.addEventListener = jest.fn();

        const menu = {
            optionOne: "Option One",
            optionTwo: "Option Two"
        };

        TestUtils.renderInWrapper(
            <DropDownButton label="Test Drop Down"
                stateless={true}
                open={true}
                options={menu}
                onToggle={jest.fn()} />
        );

        expect(TestUtils.mockCallsContains(window.addEventListener, "click")).toEqual(true);
        expect(TestUtils.mockCallsContains(window.addEventListener, "keydown")).toEqual(true);
    });

    it("unregister global listeners on unmount", function () {

        window.removeEventListener = jest.fn();

        const menu = {
            optionOne: "Option One",
            optionTwo: "Option Two"
        };

        const dropDownButtonComponent = TestUtils.renderInWrapper(
            <DropDownButton label="Test Drop Down"
                stateless={true}
                open={true}
                options={menu}
                onToggle={jest.fn()} />
        );

        //trigger unmount
        ReactDOM.unmountComponentAtNode(ReactDOM.findDOMNode(dropDownButtonComponent).parentNode);

        expect(TestUtils.mockCallsContains(window.removeEventListener, "click")).toEqual(true);
        expect(TestUtils.mockCallsContains(window.removeEventListener, "keydown")).toEqual(true);
    });

    it("triggers callback when clicked outside", function () {
        window.addEventListener = jest.fn();
        global.getSelection = jest.fn();
        global.getSelection.mockReturnValue({
            toString: () => "",
        });
        const callback = jest.fn();

        const menu = {
            optionOne: "Option One",
            optionTwo: "Option Two"
        };

        TestUtils.renderInWrapper(
            <DropDownButton label="Test Drop Down"
                stateless={true}
                open={true}
                onValueChange={jest.fn()}
                options={menu}
                onToggle={callback} />
        );

        const handler = TestUtils.findMockCall(window.addEventListener, "click")[1];


        const e = {
            target: { parentNode: document.body },
            stopPropagation: jest.fn(),
            preventDefault: jest.fn()
        };

        //click outside
        handler(e);
        expect(callback).toHaveBeenCalled();
    });

    it("doesn't trigger callback when clicked outside if not open", function () {
        window.addEventListener = jest.fn();
        const callback = jest.fn();

        const menu = {
            optionOne: "Option One",
            optionTwo: "Option Two"
        };

        TestUtils.renderInWrapper(
            <DropDownButton label="Test Drop Down"
                stateless={true}
                open={false}
                onValueChange={jest.fn()}
                options={menu}
                onToggle={callback} />
        );

        const handler = TestUtils.findMockCall(window.addEventListener, "click")[1];


        const e = {
            target: { parentNode: document.body },
            stopPropagation: jest.fn(),
            preventDefault: jest.fn()
        };

        //click outside
        handler(e);
        expect(callback).not.toHaveBeenCalled();
    });

    it("doesn't trigger callback when clicked outside and drop down is not open", function () {
        const callback = jest.fn();

        const menu = {
            optionOne: "Option One",
            optionTwo: "Option Two"
        };

        TestUtils.renderInWrapper(
            <DropDownButton label="Test Drop Down"
                stateless={true}
                open={false}
                onValueChange={jest.fn()}
                options={menu}
                onToggle={callback} />
        );

        const handler = TestUtils.findMockCall(window.addEventListener, "click")[1];
        const e = {
            target: { parentNode: document.body },
            stopPropagation: jest.fn(),
            preventDefault: jest.fn()
        };

        //click outside
        handler(e);

        expect(callback).not.toBeCalled();
    });

    it("triggers callback when drop down is open and ESC pressed", function () {

        const globalKeyListener = TestUtils.captureGlobalListener("keyDown");

        const callback = jest.fn();

        const menu = {
            optionOne: "Option One",
            optionTwo: "Option Two"
        };

        TestUtils.renderInWrapper(
            <DropDownButton label="Test Drop Down"
                stateless={true}
                open={true}
                onValueChange={jest.fn()}
                options={menu}
                onToggle={callback} />
        );

        //press ESC
        globalKeyListener({
            keyCode: 27
        });

        expect(callback).toBeCalled();
    });

    it("doesn't trigger callback when drop down is not open and ESC pressed", function () {

        const globalKeyListener = TestUtils.captureGlobalListener("keyDown");

        const callback = jest.fn();

        const menu = {
            optionOne: "Option One",
            optionTwo: "Option Two"
        };

        TestUtils.renderInWrapper(
            <DropDownButton label="Test Drop Down"
                stateless={true}
                open={false}
                onValueChange={jest.fn()}
                options={menu}
                onToggle={callback} />
        );

        //press ESC
        globalKeyListener({
            keyCode: 27
        });

        expect(callback).not.toBeCalled();
    });

    it("renders title if provided", function () {
        const component = TestUtils.renderInWrapper(
            <DropDownButton title="Test Drop Down" options={{}} label="Add" />
        );

        // no title in collapsed mode
        const titles = TestUtils.scryRenderedDOMNodesWithDataId(component, "options-title");
        expect(titles.length).toBe(0);

        // toggle the drop down
        const toggle = TestUtils.findRenderedDOMNodeWithDataId(component, "action");
        ReactTestUtils.Simulate.click(toggle);

        const title = TestUtils.findRenderedDOMNodeWithDataId(component, "options-title");
        expect(title.textContent).toBe("Test Drop Down");
    });

    it("renders classname and data-id", function () {
        const component = TestUtils.renderInWrapper(
            <DropDownButton options={{}} label="Add" data-id="my-id" className="my-class" />
        );

        const node = ReactDOM.findDOMNode(component);

        expect(node.getAttribute("data-id")).toBe("my-id");
        expect(node.getAttribute("class")).toContain("my-class");
    });

    it("render component with default data-id", function () {
        const component = TestUtils.renderInWrapper(
            <DropDownButton options={{}} label="Add" />
        );

        const element = TestUtils.findRenderedDOMNodeWithDataId(component, "drop-down-button");
        expect(element).toBeDefined();
    });

    it("renderButton renders correct content and passes toggle function", () => {
        const renderButton = ({
            onClick
        }) => <button data-id="render-button" onClick={onClick} />;

        const component = TestUtils.renderInWrapper(
            <DropDownButton
                label="Add"
                options={{
                    one: "one"
                }}
                renderButton={renderButton}
            />
        );

        const button = TestUtils.findRenderedDOMNodeWithDataId(component, "render-button");
        expect(button).toBeTruthy();

        ReactTestUtils.Simulate.click(button);

        const option = TestUtils.findRenderedDOMNodeWithDataId(component, "option-one");
        expect(option).toBeTruthy();
    });

    it("correctly opens and closes", () => {
        const component = mount(
            <DropDownButton
                label="Label"
                options={{
                    one: "one"
                }}
            />
        );

        expect(component.find(".dropdown-button__options").exists()).toEqual(false);

        component.find("[data-id=\"action\"]").simulate("click");

        expect(component.find(".dropdown-button__options").exists()).toEqual(true);
    });
});
