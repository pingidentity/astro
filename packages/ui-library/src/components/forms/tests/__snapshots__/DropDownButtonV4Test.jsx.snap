// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropDownButton data-id's don't change 1`] = `
Array [
  "drop-down-button",
  "drop-down-button-trigger",
  "action",
]
`;
